### PR TITLE
[SPARK-32542][SQL] Add an optimizer rule to split an Expand into multiple Expands for aggregates

### PR DESCRIPTION
**Pull request by Karl-WangSK**
### What changes were proposed in this pull request?
Add an optimizer rule to split an Expand into multiple Expands for aggregates

### Why are the changes needed?

Split an expand into several small Expand, which contains the Specified number of projections.
For instance, like this sql.select a, b, c, d, count(1) from table1 group by a, b, c, d with cube. It can expand 2^4 times of original data size.
Now we specify the spark.sql.optimizer.projections.size=4.The Expand will be split into 2^4/4 smallExpand.It can reduce the shuffle pressure and improve performance in multidimensional analysis when data is huge.

```
runBenchmark("cube multianalysis agg") {
      val N = 20 << 20

      val benchmark = new Benchmark("cube multianalysis agg", N, output = output)

      def f(): Unit = {
        val df = spark.range(N).cache()
        df.selectExpr(
          "id",
          "(id & 1023) as k1",
          "cast(id & 1023 as string) as k2",
          "cast(id & 1023 as int) as k3",
          "cast(id & 1023 as double) as k4",
          "cast(id & 1023 as float) as k5")
          .cube("k1", "k2", "k3", "k4", "k5")
          .sum()
          .noop()
        df.unpersist()

      }

      benchmark.addCase("grouping = F") { _ =>
        withSQLConf(
          SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true",
          SQLConf.GROUPING_WITH_UNION.key -> "false") {
          f()
        }
      }

      benchmark.addCase("grouping = T projectionSize= 16") { _ =>
        withSQLConf(
          SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true",
          SQLConf.GROUPING_WITH_UNION.key -> "true",
          SQLConf.GROUPING_EXPAND_PROJECTIONS.key -> "16") {
          f()
        }
      }

      benchmark.addCase("grouping = T projectionSize= 8") { _ =>
        withSQLConf(
          SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true",
          SQLConf.GROUPING_WITH_UNION.key -> "true",
          SQLConf.GROUPING_EXPAND_PROJECTIONS.key -> "8") {
          f()
        }
      }
      benchmark.run()
    }
```

```
Running benchmark: cube multianalysis agg :
cube 5 fields k1, k2, k3, k4, k5
Running case: GROUPING_WITH_UNION off 
Running case: GROUPING_WITH_UNION on and GROUPING_EXPAND_PROJECTIONS = 16 
Running case: GROUPING_WITH_UNION on and GROUPING_EXPAND_PROJECTIONS = 8

Java HotSpot(TM) 64-Bit Server VM 1.8.0_181-b13 on Mac OS X 10.15
Intel(R) Core(TM) i5-7267U CPU @ 3.10GHz
cube multianalysis agg:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
grouping = F                                      54329          54931         852          0.4        2590.6       1.0X
grouping = T projectionSize= 16                   44584          44781         278          0.5        2125.9       1.2X
grouping = T projectionSize= 8                    42764          43272         718          0.5        2039.1       1.3X
```

```
Running benchmark: cube multianalysis agg :
cube 6 fields k1, k2, k3, k4, k5, k6
Running case: GROUPING_WITH_UNION off 
Running case: GROUPING_WITH_UNION on and GROUPING_EXPAND_PROJECTIONS = 32
Running case: GROUPING_WITH_UNION on and GROUPING_EXPAND_PROJECTIONS = 16

Java HotSpot(TM) 64-Bit Server VM 1.8.0_181-b13 on Mac OS X 10.15
Intel(R) Core(TM) i5-7267U CPU @ 3.10GHz
cube multianalysis agg:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
grouping = F                                     141607         143424        2569          0.1        6752.4       1.0X
grouping = T projectionSize= 32                  109465         109603         196          0.2        5219.7       1.3X
grouping = T projectionSize= 16                   99752         100411         933          0.2        4756.5       1.4X
```

```
Running benchmark: cube multianalysis agg :
cube 7 fields k1, k2, k3, k4, k5, k6, k7
Running case: GROUPING_WITH_UNION off 
Running case: GROUPING_WITH_UNION on and GROUPING_EXPAND_PROJECTIONS = 64
Running case: GROUPING_WITH_UNION on and GROUPING_EXPAND_PROJECTIONS = 32
Running case: GROUPING_WITH_UNION on and GROUPING_EXPAND_PROJECTIONS = 16

Java HotSpot(TM) 64-Bit Server VM 1.8.0_181-b13 on Mac OS X 10.15
Intel(R) Core(TM) i5-7267U CPU @ 3.10GHz
cube multianalysis agg:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
grouping = F                                     516941         519658         NaN          0.0       24649.7       1.0X
grouping = T projectionSize= 64                  267170         267547         533          0.1       12739.6       1.9X
grouping = T projectionSize= 32                  252023         255307        1769          0.1       12017.4       2.1X
grouping = T projectionSize= 16                  235929         236944        1436          0.1       11250.0       2.2X
```

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
unit test

### CODE REVIEW

**Comment by maropu** [comment]
Thanks for the contribution, @Karl-WangSK. btw, could you show us performance numbers with/without this fix in the PR description first?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -2690,6 +2690,19 @@ object SQLConf {
...
+    .doc("When true,enable grouping with union.")
+    .booleanConf
+    .createWithDefault(false)
+
+  val GROUPING_EXPAND_PROJECTIONS = buildConf("spark.sql.optimizer.projections.size")
+    .doc("Split a expand into several smallExpand,which contains the size of projections.It can" +
+      "reduce the shuffle pressure and improve performance in multidimensional analysis " +
+      "when data is huge.")
+    .intConf
+    .createWithDefault(4)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2704]: Could you check if all the existing tests pass if `spark.sql.optimizer.grouping`=true and `spark.sql.optimizer.projections.size`=1?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -1823,3 +1824,33 @@ object OptimizeLimitZero extends Rule[LogicalPlan] {
       empty(ll)
   }
 }
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1827]: nit: remove this single blank.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -2699,6 +2712,7 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2715]: nit: remove this.

**Comment by maropu** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (Karl-WangSK) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ok!

**Comment by maropu** [comment]
@Karl-WangSK Why did you close this? If you gave up working on this, could we close the correspoinding jira, too?

**Comment by Karl-WangSK** [comment]
> @Karl-WangSK Why did you close this? If you gave up working on this, could we close the correspoinding jira, too?

sry for closing this pr. sql test failed. didn't find the reason. need time to fix the bugs.

**Comment by maropu** [comment]
If you're still working on this, its okay to keep this open.

**Comment by maropu** [comment]
btw, could you make the PR title clearer? I think the current one looks ambigous.

**Comment by maropu** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -2690,6 +2690,18 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val GROUPING_WITH_UNION = buildConf("spark.sql.optimizer.grouping")
+    .doc("When true,enable grouping with union.")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2694]: For text style. Please add a space after punctuation marks like , or .

For example, "When true,enable grouping with union." -> "When true, enable grouping with union."

There are several places with this style issue.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -1823,3 +1824,32 @@ object OptimizeLimitZero extends Rule[LogicalPlan] {
...
+/**
+ * Split [[Expand]] into several splitExpand if the projection size of Expand is larger
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1829]: I think you mean "Split [[Expand]] into several [[Expand]] ...". There isn't a class called `splitExpand`.

Comment (Karl-WangSK) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -2690,6 +2690,18 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val GROUPING_WITH_UNION = buildConf("spark.sql.optimizer.grouping")
+    .doc("When true, enable grouping with union.")
+    .booleanConf
+    .createWithDefault(false)
+
+  val GROUPING_EXPAND_PROJECTIONS = buildConf("spark.sql.optimizer.projections.size")
+    .doc("Split a expand into several smallExpand, which contains the size of projections. It " +
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2699]: expand -> Expand

smallExpand -> small Expand

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 7]
```diff
@@ -2690,6 +2690,18 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val GROUPING_WITH_UNION = buildConf("spark.sql.optimizer.grouping")
+    .doc("When true, enable grouping with union.")
+    .booleanConf
+    .createWithDefault(false)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2696]: Is this config needed?

Comment (Karl-WangSK) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2696]: yes. The default spark.sql.optimizer.projections.size=4,but if your sql's projections excceed 4,and you still don't want to turn on this function. you can set spark.sql.optimizer.grouping=false

**Review by viirya — COMMENTED** [review]
Can you add some tests for this? And, as @maropu said, please also run some benchmarks to show the performance improvement. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 8]
```diff
@@ -1823,3 +1824,32 @@ object OptimizeLimitZero extends Rule[LogicalPlan] {
...
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case a @ Aggregate(_, _, e @ Expand(projections, _, _)) =>
+      if (SQLConf.get.groupingWithUnion && projections.length
+        > SQLConf.get.groupingExpandProjections) {
+        val num = SQLConf.get.groupingExpandProjections
+        val subExpands = splitExpand(e, num)
+        val aggregates: Seq[Aggregate] = subExpands.map { expand =>
+          Aggregate(a.groupingExpressions, a.aggregateExpressions, expand)
+        }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1849]: These split expand + aggregate read same data. That's said this optimization will read same data multiple times.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 9]
```diff
@@ -1823,3 +1824,32 @@ object OptimizeLimitZero extends Rule[LogicalPlan] {
       empty(ll)
   }
 }
+
+/**
+ * Split [[Expand]] into several Expand if the projection size of Expand is larger
+ * than default projection size.
+ */
+object SplitAggregateWithExpand extends Rule[LogicalPlan] {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1832]: I'm curious about how this reduces shuffled data. I think the original idea is, in the cases of Expand will produce huge data, splitting it to small ones.

But shuffle is happened during Aggregate here, right? By splitting, the total amount of shuffled data is not changed, but split into several ones. Does it really result significant improvement?

I'm curious to look at the benchmark results.

**Comment by Karl-WangSK** [comment]
> Can you add some tests for this? And, as @maropu said, please also run some benchmarks to show the performance improvement. Thanks.

ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 8 above)
Comment (Karl-WangSK) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: not the same data, they have passed the sql tests, the output is the same as expected

**Comment by Karl-WangSK** [comment]
retest this please

**Comment by Karl-WangSK** [comment]
retest this please

**Comment by Karl-WangSK** [comment]
@viirya  @maropu

**Comment by Karl-WangSK** [comment]
please review the benchmark

**Comment by maropu** [comment]
> But shuffle is happened during Aggregate here, right? By splitting, the total amount of shuffled data is not changed, but split into several ones. Does it really result significant improvement?

As @viirya said above, I think the same. Why can this reduce the amount of shuffle writes? In the case of `expand -> partial aggregates`, the aggregates seem to have the same **total** amount of output size.

**Comment by Karl-WangSK** [comment]
yes.The shuffle output is the same, because the size of the data is the same. 
As you can see the benchmark:
cube 7 fields k1, k2, k3, k4, k5, k6, k7(128x projections)  and cube 6 fields k1, k2, k3, k4, k5, k6(64x projections) with  grouping off
data size is double ,but the time ,one is 2.4min ,the another one is 8.7min, not just double time .It will be affected by data size Especially when the memory is limited.
The original data I created is about 20M, executor memory is 1g. when it expands to 64x or  128x. It will have big impact on shuffle performance.

**Comment by Karl-WangSK** [comment]
hi. Can anyone deal with this pr?

**Comment by Karl-WangSK** [comment]
@cloud-fan

**Comment by Karl-WangSK** [comment]
retest this please

**Comment by Karl-WangSK** [comment]
@cloud-fan @maropu  Please take a look, thanks!

**Comment by Karl-WangSK** [comment]
> > But shuffle is happened during Aggregate here, right? By splitting, the total amount of shuffled data is not changed, but split into several ones. Does it really result significant improvement?
> 
> As @viirya said above, I think the same. Why can this reduce the amount of shuffle writes (and improve the performance)? In the case of `expand -> partial aggregates`, the aggregates seem to have the same **total** amount of output size.

In my idea and according to benchmark.I think the when data size is larger than execution memory ,then it will cache in disk ,so it losses the performance and increase the time.

**Comment by Karl-WangSK** [comment]
retest this please

**Comment by Karl-WangSK** [comment]
ready to merge if no other problems @LuciferYang  Thanks!

**Pull request closed without merging**